alf/environments/carla_sensors.py

```diff
@@ -1588,87 +1640,162 @@ def is_within_distance(loc):
+            for i, mask in enumerate(mask_dict["walker_masks"]):
+                image[:, mask] = interp_color(COLOR_CYAN, (h_len - i) * 0.2)
+
+            image[:, mask_dict["ev_mask"]] = COLOR_WHITE
+
+            return image
+
+        elif self._observation_mode == "mask":
             # use masks
-            c_road = road_mask * 255
-            c_route = route_mask * 255
-            c_lane = lane_mask * 255
+            c_road = mask_dict["road_mask"] * 255
+            c_route = mask_dict["route_mask"] * 255
+            c_lane = mask_dict["lane_mask"] * 255
 
             # masks with history
+            # [road, route, lane, [vehicle, walker, tl_g_y_r] * hist_len]
             c_tl_history = []
             for i in range(len(self._history_idx)):
-                c_tl = np.zeros([self._width, self._width], dtype=np.uint8)
-                c_tl[tl_green_masks[i]] = 80
-                c_tl[tl_yellow_masks[i]] = 170
-                c_tl[tl_red_masks[i]] = 255
-                c_tl[stop_masks[i]] = 255
+                c_tl = np.zeros([self._height, self._width], dtype=np.uint8)
+                c_tl[mask_dict["tl_green_masks"][i]] = 80
+                c_tl[mask_dict["tl_yellow_masks"][i]] = 170
+                c_tl[mask_dict["tl_red_masks"][i]] = 255
+                c_tl[mask_dict["stop_masks"][i]] = 255
                 c_tl_history.append(c_tl)
 
-            c_vehicle_history = [m * 255 for m in vehicle_masks]
-            c_walker_history = [m * 255 for m in walker_masks]
+            c_vehicle_history = [m * 255 for m in mask_dict["vehicle_masks"]]
+            c_walker_history = [m * 255 for m in mask_dict["walker_masks"]]
 
             masks = np.stack((c_road, c_route, c_lane, *c_vehicle_history,
                               *c_walker_history, *c_tl_history),
-                             axis=2)
-            masks = np.transpose(masks, [2, 0, 1])
-            return masks
+                             axis=0)
+
+            return masks.astype(np.uint8)
+
+        elif self._observation_mode == "bitmap":
+            # use masks
+            c_road = mask_dict["road_mask"]
+            c_route = mask_dict["route_mask"]
+            c_lane = mask_dict["lane_mask"]
+
+            # masks with history
+            # [road, route, lane, [vehicle, walker, tl_g, tl_y, tl_r] * hist_len]
+            c_tl_g_history = []
+            c_tl_y_history = []
+            c_tl_r_history = []
+            c_vehicle_history = []
+            c_walker_history = []
+            for i in range(len(self._history_idx)):
+                c_tl_g_history.append(mask_dict["tl_green_masks"][i])
+                c_tl_y_history.append(mask_dict["tl_yellow_masks"][i])
+                c_tl_r_history.append(mask_dict["tl_red_masks"][i])
+
+            c_vehicle_history = [m * 255 for m in mask_dict["vehicle_masks"]]
+            c_walker_history = [m * 255 for m in mask_dict["walker_masks"]]
```

Comment:
I am a bit confused by those variables as they seemed to be unused below. Are they actually useful?

Reply:
> I am a bit confused by those variables as they seemed to be unused below. Are they actually useful?

Sorry about the confusion. These are some obsoleted code that are not used anymore. I have removed them.